libcsm/os.py

```diff
@@ -86,9 +86,15 @@ def _run(self) -> None:
             self._return_code = error.errno
             LOG.error('Could not find command for given args: %s', self.args)
         else:
-            self._stdout = stdout.decode('utf8')
-            self._stderr = stderr.decode('utf8')
-            self._return_code = command.returncode
+            try:
+                self._stdout = stdout.decode('utf8')
+                self._stderr = stderr.decode('utf8')
```

Comment:
Why are we forcing utf8 decoding for all output from commands we wrap? If any of those commands output vt100 escape sequences as an example this of course will fail but thats not unexpected, the output is legitimately not utf8, our expectations of this byte stream being utf8 is... not going to survive in many LANG or LOCALE's python could be running in.

What problem are we trying to solve exactly here?

Reply:
We can auto-detect the encoding with something like.

```python
import chardet
result = chardet.detect('string')
encoding = result['encoding']
```

Reply:
True, but for a general purpose wrapper, we shouldn't be decoding command output at all. All any unix command guarantees is it will output a stream of bytes. I could encode a file name as ISO-8859 and write that out to the filesystem and ls -l on that file will always fail a utf8 decode, because well it is invalid and always will be. Utf decoding should probably be an optional kwarg that is off by default would be a better api for this and let callers decide if the command they wanted to run needs to truly have utf8 output or not.

Reply:
And one problem with detection is we can always have mixed encoding in the same stream,  eg one surprise result could be this (yes i used german for surprise to be cheeky):
```
>>> result = chardet.detect("Überraschung".encode('cp1252') + "Überraschung".encode('utf8'))
>>> encoding = result['encoding']
>>> print(result)
{'encoding': 'Windows-1254', 'confidence': 0.6321134231346641, 'language': 'Turkish'}
```

Reply:
Ok, it sounds like you're suggesting we just remove the decode and then the caller of `run_command` is responsible for decoding (if at all) the bytes in stdout and stderr.

I can add an optional param to decode for a given encoding.

Reply:
Yep basically that, have the auto encode to utf-8 be a deliberate choice call sites opt into. That then is the contract that the call site should deal with any encoding errors if any.

The other reason the call site is the "right"er spot for this exception is we don't want to just dump the stdout/error of the command that was run as we're then losing data just because we couldn't decode it to utf8. That really seems a horrible api for "run a command and lose the output entirely if it can't be decoded to utf8 due to the output it can't control anyway".

Having it be optional and opt-in then at least means the call site can catch this exception when they opt into it but I don't like losing the commands output at all even in that case but if someone opted into it the presumption can be they're ok with losing the commands output in a decode error or not. But we probably should think a bit more about what contracts we're basically building when we add these wrappers around things like popen.